syntax = "proto3";

package aafs;

message Stat {
  uint64 ino = 1;   /* Inode number */
  uint64 mode = 2;  /* File type and mode */
  uint64 nlink = 3; /* Number of hard links */
  uint64 uid = 4;   /* User ID of owner */
  uint64 gid = 5;   /* Group ID of owner */
  uint64 rdev = 6;  /* Device ID (if special file) */
  int64 size = 7;   /* Total size, in bytes */
  int64 blocks = 8; /* Number of 512B blocks allocated */
  int64 atime = 9;
  int64 mtime = 10;
  int64 ctime = 11;
}

message PathRequest { string path = 1; }

message GetAttrResponse {
  int32 ret = 1;
  Stat stat = 2;
}

message ReadDirResponse {
  int32 ret = 1;
  repeated string entries = 2;
}

message FileContent { bytes data = 1; }

service gRPCService {
  rpc s_getattr(PathRequest) returns (GetAttrResponse) {}
  rpc s_readdir(PathRequest) returns (ReadDirResponse) {}
  rpc s_download(PathRequest) returns (stream FileContent) {}
}